rust/src/builtins/scriptlet_intercept/groupadd.rs

```diff
@@ -0,0 +1,150 @@
+//! CLI handler for intercepted `groupadd`.
+
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use anyhow::Result;
+use cap_std::fs::Dir;
+use clap::{Arg, Command};
+use std::io::Write;
+
+/// Entrypoint for (the rpm-ostree implementation of) `groupadd`.
+pub(crate) fn entrypoint(args: &[&str]) -> Result<()> {
+    fail::fail_point!("intercept_groupadd_ok", |_| Ok(()));
+
+    // This parses the same CLI surface as the real `groupadd`,
+    // but in the end we only extract the group name and (if
+    // present) the static GID.
+    let matches = cli_cmd().get_matches_from(args);
+    let gid: Option<u32> = match matches.value_of("gid") {
+        None => None,
+        Some(s) => Some(s.parse()?),
+    };
+    let groupname = matches
+        .value_of("groupname")
+        .expect("missing required groupname");
+
+    let rootdir = Dir::open_ambient_dir("/", cap_std::ambient_authority())?;
+    generate_sysusers_fragment(&rootdir, groupname, gid)?;
+
+    Ok(())
+}
+
+/// CLI parser, matches <https://linux.die.net/man/8/groupadd>.
+fn cli_cmd() -> Command<'static> {
+    let name = "groupadd";
+    Command::new(name)
+        .bin_name(name)
+        .about("create a new group")
+        .arg(Arg::new("force").short('f').long("force"))
+        .arg(
+            Arg::new("help")
+                .short('h')
+                .long("help")
+                .action(clap::ArgAction::Help),
+        )
+        .arg(Arg::new("gid").short('g').long("gid").takes_value(true))
+        .arg(Arg::new("key").short('K').long("key").takes_value(true))
+        .arg(Arg::new("allow_duplicates").short('o').long("non-unique"))
+        .arg(
+            Arg::new("password")
+                .short('p')
+                .long("password")
+                .takes_value(true),
+        )
+        .arg(Arg::new("system").short('r').long("system"))
+        .arg(
+            Arg::new("chroot_dir")
+                .short('R')
+                .long("root")
+                .takes_value(true),
+        )
+        .arg(
+            Arg::new("prefix_dir")
+                .short('P')
+                .long("prefix")
+                .takes_value(true),
+        )
+        .arg(Arg::new("users").short('U').long("users").takes_value(true))
+        .arg(Arg::new("groupname").required(true))
+}
+
+/// Write a sysusers.d configuration fragment for the given group.
+///
+/// This returns whether a new fragment has been actually written
+/// to disk.
+fn generate_sysusers_fragment(rootdir: &Dir, groupname: &str, gid: Option<u32>) -> Result<bool> {
+    static SYSUSERS_DIR: &str = "usr/lib/sysusers.d";
+
+    // The filename of the configuration fragment is in fact a public
+    // API, because users may have masked it in /etc. Do not change this.
+    let filename = format!("30-pkg-group-{groupname}.conf");
```

Comment:
For tmpfiles snippets we derive the filename from the package name - I think we could do that here by passing it through the environment.

I don't have a strong opinion on that though.

Reply:
I considered doing that, I explored the pkgname-through-env approach and it's indeed feasible.

I'm still oscillating on the file-naming decision though.
My current stance is that having `{groupname}` in the file helps with idempotence, as a single package may be doing multiple `groupadd`. Current naming convention avoid interferences and allows for atomic-replaces without parsing/merging.
One third option I considered is the `{pkgname}-{groupname}` format, but I didn't see many additional benefits in doing that.

I'm open to more brainstorming on this topic. /cc @jlebon.

Reply:
Hmm, looking at `/usr/lib/sysusers.d` in my local FSB and FCOS, I don't see any config using priority prefixes in their names (except Zincati). So even with `30-`, we'd be at the top of the list.

>  One third option I considered is the `{pkgname}-{groupname}` format, but I didn't see many additional benefits in doing that.

That'd be my vote. I think the additional benefit of being able to tell which package we derived from is nice (given that we don't currently include the package name in the generated fragment either). It would also better conform to the guideline established in `sysusers.d(5)`:

> Each configuration file shall be named in the style of `package.conf` or `package-part.conf`. The second variant should be used when it is desirable to make it easy to override just this part of configuration.

(Though it's a bit fuzzy here because we're doing it for them, hehe, but it's in the same spirit I think).

Reply:
Ok, it looks like there is consensus about adding the package name in the filename. Sorry I didn't see the last comment here before rebasing, I'll wire the pkgname in a followup.

Regarding the numerical prefix and suggested naming pattern, there has been some friction reported upstream already: https://github.com/systemd/systemd/issues/21155

In our case, I already know we'd need to properly priority-sort several sources with ([possibly conflicting](https://github.com/coreos/fedora-coreos-tracker/issues/1201)) definitions:
 * packaged fragment (e.g. https://pagure.io/setup/pull-request/39)
 * file overlays (e.g. https://github.com/coreos/fedora-coreos-config/pull/1836)
 * `groupadd` translation (this PR)

While we aren't strictly following the suggested pattern from the docs, the numerical prefix at least allows predictable sorting/overriding which is helpful in the above case of multiple/conflicting sources of truth.